feat: custom chart colors support within existing `colors` `prop`

adamhenson opened the pull request:
**Description**

This change is refactor of #836 to provide support for custom colors, but instead of the additive `customChartColors` `prop` - this ports functionality to the existing `colors` `prop`.

With this change, the `colors` `prop` can accept custom colors or tremor base colors (`(Color | string)[]`). `customChartColors` is removed in this change.

**Example Usage**

```tsx
<LineChart
  {...props}
  colors={[
    // now supports any mix of tremor base colors...
    'emerald',
    'indigo',

    // ... and custom colors
    '#32a852',
    'custom-color-100',
  ]}
/>
```

Example `safelist` needed in Tailwind config to support the above implementation:

```typescript
safelist: [
  // ..
  // ..
  // custom colors
  ...["[#32a852]", "custom-color-100"].flatMap((customColor) => [
    `bg-${customColor}`,
    `border-${customColor}`,
    `hover:bg-${customColor}`,
    `hover:border-${customColor}`,
    `hover:text-${customColor}`,
    `fill-${customColor}`,
    `ring-${customColor}`,
    `stroke-${customColor}`,
    `text-${customColor}`,
    `ui-selected:bg-${customColor}`,
    `ui-selected:border-${customColor}`,
    `ui-selected:text-${customColor}`,
  ]),
],
```

**Updated Components**

- `AreaChart`
- `BarChart`
- `DonutChart`
- `LineChart`
- `ScatterChart`
- `SparkAreaChart`
- `SparkBarChart`
- `SparkLineChart`

**Related issues and PRs**

- Related #836
- Related #565 
- Related #80 
- Related #640 

**What kind of change does this PR introduce?**

Although this change updates the existing `colors` `prop`, the functionality is additive and doesn't break existing behavior.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**How has This been tested?**

Storybook stories have been added for all components demonstrating new support of custom colors in the `colors` `prop`.

**The PR fulfills these requirements:**

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] It's submitted to the `main` branch
- [x] When resolving a specific issue, it's referenced in the related issue section above
- [x] My change requires a change to the documentation. (Managed by Tremor Team)
- [x] I have added tests to cover my changes
- [x] Check the ["Allow edits from maintainers" option](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) while creating your PR.
- [x] Add refs #XXX or fixes #XXX to the related issue section if your PR refers to or fixes an issue.
- [x] By contributing to Tremor, you confirm that you have read and agreed to Tremor's [CONTRIBUTING.md](https://github.com/tremorlabs/tremor/blob/main/CONTRIBUTING.md) guideline. You also agree that your contributions will be licensed under the [Apache License 2.0](https://github.com/tremorlabs/tremor/blob/main/License) license.

severinlandolt commented:
This PR is what open-source maintainers dream of🫶 Will go to bed now and review it tomorrow!

adamhenson commented:
Hah! Awesome, thanks @severinlandolt - appreciated 🙏

severinlandolt commented:
Just released a beta and testing it on our website. Two small questions:

1) Legend
I think we do not need the customColor prop in the svg anymore, or am I missing something?
![CleanShot 2023-12-10 at 13 49 07@2x](https://github.com/tremorlabs/tremor/assets/41927988/cb428589-a489-4b50-a9a0-df45d87bd6e2)

2) Interestingly, I use Tremor together with MDXRemote, and there is a TS error popping up. I am not 100% sure if this is a problem of Tremor or MDXRemote. I will do some tests, but I suppose there should be a type for the `colorValues` array? 
`export const colorValues: string[] = [...`
![CleanShot 2023-12-10 at 13 52 12@2x](https://github.com/tremorlabs/tremor/assets/41927988/33dfd976-5503-4534-bcab-134bdcff2ef2)
![CleanShot 2023-12-10 at 13 54 49@2x](https://github.com/tremorlabs/tremor/assets/41927988/253996d2-7b86-484d-8cba-b30c90b4e47a)

adamhenson commented:
Thanks @severinlandolt! And thanks for testing. Yes, correct, we don't need `customColor`... I must have missed that one. Regarding the second issue - I wonder if there is a way I can reproduce 🤔 

In the meantime, I'll open a branch against `beta-customcolors` to fix that first issue.

adamhenson commented:
Would you be able to provide the full error @severinlandolt?

severinlandolt commented:
Sure @adamhenson,

```
Type '{ Accordion: ForwardRefExoticComponent<tremor.AccordionProps & RefAttributes<HTMLDivElement>>; AccordionBody: ForwardRefExoticComponent<...>; ... 180 more ...; blockquote: (props: any) => JSX.Element; }' is not assignable to type 'MDXComponents | MergeComponents | null | undefined'.
  Type '{ Accordion: ForwardRefExoticComponent<tremor.AccordionProps & RefAttributes<HTMLDivElement>>; AccordionBody: ForwardRefExoticComponent<...>; ... 180 more ...; blockquote: (props: any) => JSX.Element; }' is not assignable to type 'MDXComponents'.
    Type '{ Accordion: ForwardRefExoticComponent<tremor.AccordionProps & RefAttributes<HTMLDivElement>>; AccordionBody: ForwardRefExoticComponent<...>; ... 180 more ...; blockquote: (props: any) => JSX.Element; }' is not assignable to type 'NestedMDXComponents'.
      Property 'colorValues' is incompatible with index signature.
        Type 'readonly ["slate", "gray", "zinc", "neutral", "stone", "red", "orange", "amber", "yellow", "lime", "green", "emerald", "teal", "cyan", "sky", "blue", "indigo", "violet", "purple", "fuchsia", "pink", "rose"]' is not assignable to type 'NestedMDXComponents | Component<any>'.
          Type 'readonly ["slate", "gray", "zinc", "neutral", "stone", "red", "orange", "amber", "yellow", "lime", "green", "emerald", "teal", "cyan", "sky", "blue", "indigo", "violet", "purple", "fuchsia", "pink", "rose"]' is not assignable to type 'NestedMDXComponents'.
            Index signature for type 'string' is missing in type 'readonly ["slate", "gray", "zinc", "neutral", "stone", "red", "orange", "amber", "yellow", "lime", "green", "emerald", "teal", "cyan", "sky", "blue", "indigo", "violet", "purple", "fuchsia", "pink", "rose"]'.ts(2322)
index.d.ts(30, 5): The expected type comes from property 'components' which is declared here on type 'IntrinsicAttributes & MDXRemoteSerializeResult<Record<string, unknown>, Record<string, unknown>> & { components?: MDXComponents | MergeComponents | null | undefined; lazy?: boolean | undefined; }'
```

I try to create a small repo to replicate the issue in the meantime

adamhenson commented:
Also, with that said this could be a breaking change in types if `Color` type is being imported and expected to be `colors: Color[]`. I could see indexing being a problem. I am surprised the `colorValues: readonly string[]` change didn't work. But still digging in.

I'm wondering if we should do something like this and replace all occurrences of `(Color | string)[]`.

```diff
- export type Color = (typeof colorValues)[number];
+ export type Color = (typeof colorValues)[number] | string;
```

severinlandolt commented:
Hey @adamhenson, I added you as an outside collaborator to this repo so feel free to directly push to https://github.com/tremorlabs/tremor/tree/beta-customcolors

adamhenson commented:
Great, thanks @severinlandolt!

adamhenson commented:
Another thought I had is maybe the build tool doesn't like my type casting in this one.

```typescript
export const getIsBaseColor = (color: Color | string) => colorValues.includes(color as Color);
```

But will wait to look at the repro.

severinlandolt commented:
Alright, sent you another invitation 📨

adamhenson commented:
Perfect, thanks @severinlandolt 👀

severinlandolt commented:
@adamhenson Feel free to make as many beta releases of https://github.com/tremorlabs/tremor/tree/beta-customcolors as you wish. Just prefix your commit with `fix:` and a new beta should be released

adamhenson commented:
Oh awesome, thanks @severinlandolt! I might be able to get away with local testing via `npm link` but if not, that will be super helpful.

adamhenson commented:
I have a theory that it's because I added `export` to export `colorValues` and it looks like it's being exported from the main package and being interpreted as a component. About to test this now.

From the built `index.d.ts` from the `tremor` package.

<img width="1181" alt="Screenshot 2023-12-10 at 12 36 25 PM" src="https://github.com/tremorlabs/tremor/assets/747743/a542d03b-6a56-4f48-b8e4-26c14015f792">

I can remove that `export` statement anyways, since I'm no longer using it in other files. My theory is that this will fix it. More coming soon... just need to test.

severinlandolt commented:
True, in the previous version the type in question was not exported!

adamhenson commented:
Yep! It does fix it but, now I see a different build error... but seems unrelated?

```
./data/blocks/components/landing-cards/LandingCardListBar.tsx:152:16
Type error: Type 'VisitsData[]' is not assignable to type 'Bar[]'.
  Type 'VisitsData' is not assignable to type 'Bar'.
    Types of property 'icon' are incompatible.
      Type 'JSXElementConstructor<any>' is not assignable to type 'JSXElementConstructor<any> | undefined'.
        Type '(props: any, deprecatedLegacyContext?: any) => ReactNode' is not assignable to type 'JSXElementConstructor<any> | undefined'.
          Type '(props: any, deprecatedLegacyContext?: any) => React.ReactNode' is not assignable to type '(props: any, deprecatedLegacyContext?: any) => import("/Users/adam/Projects/tremor/node_modules/@types/react/index").ReactNode'.
            Type 'React.ReactNode' is not assignable to type 'import("/Users/adam/Projects/tremor/node_modules/@types/react/index").ReactNode'.
              Type 'ReactElement<any, string | JSXElementConstructor<any>>' is not assignable to type 'ReactNode'.
                Property 'children' is missing in type 'ReactElement<any, string | JSXElementConstructor<any>>' but required in type 'ReactPortal'.

  150 |         </Text>
  151 |       </Flex>
> 152 |       <BarList data={filteredData} showAnimation={false} className="mt-4" />
      |                ^
  153 |     </Card>
  154 |   );
  155 | }
```

adamhenson commented:
@severinlandolt, yep - `3.12.0-beta-customcolors.5` fixes! I just removed the `export` - it was no longer needed anyways. I deleted my previous comment because that error might have been a bi-product of `npm link`. I ran the build with the new release and build completes successfully. I pushed to the bug repo too with the updated version.

adamhenson commented:
Phew, thanks for all the help here @severinlandolt.

severinlandolt commented:
No, I have to thank you! I will test the new beta in my repos asap🚀

adamhenson commented:
Hah! Yep, the main entrypoint of the package exports everything from `./lib/inputTypes`.

https://github.com/tremorlabs/tremor/blob/1744e4dda4562d17e74947c5e2804578677177ed/src/index.ts#L2

And the repro project expects all those imports to be components.

adamhenson commented:
@severinlandolt Just saw the big news—Tremor + Vercel! Huge congrats! 🎉 Super exciting to see where this goes, especially with v0 and generative UI.

Loved contributing to the first iteration of Tremor (still honored by the "open-source maintainer’s dream" comment 😆). If you ever need a UI nerd at Vercel, I’m just saying… I do have a very particular set of skills. 😉

Let’s connect! — [My LinkedIn](https://www.linkedin.com/in/adamhenson1979/)
